### DISCUSSION TITLE
How to manually poll Runtime to drive futures without blocking current thread?

### DISCUSSION THREAD

**Discussion by alshdavid** [tokio-rs/tokio#6651] [Q&A]
Hey all, I am trying to bolt tokio onto an external event loop (libuv via a Nodejs addon). 

For this, I need to enter tokio and manually drive futures to completion using a current thread scheduler which does not block if there are pending futures.

Essentially, I need it to loop over all of the internal futures, poll them, continuing on pending futures.

Currently tokio only has the `block_on` method which blocks the main thread (consequently blocking the external event loop). I essentially need a `try_next_tick()` method.

I've been trying to reconstruct the `CurrentThread` scheduler in user code but almost all of the internals required to do this are private to the tokio crate - and I don't want to fork Tokio.

Is there an obvious way to achieve this that I am missing? 

I was able to achieve this successfully with `smol::LocalExecutor::try_tick()` but I'd rather use Tokio.

Trying to get this working with Tokio in a sandbox I am trying the following:

```rust
fn main() {
  let runtime = Builder::new_current_thread().enable_all().build().unwrap();

  let mut handle = runtime.spawn(async {
    println!("1");
    tokio::time::sleep(Duration::from_secs(1)).await;
    println!("2");
  });

  let _ = runtime.enter();

  // external event loop
  loop { 
    poll::poll_fn(|ctx| {
      Future::poll(Pin::new(&mut handle), ctx)
    });
  }
}
```

I have tried using a `LocalSet` but I figure if I can run a single future like this, I can run a `LocalSet` like this so I am starting simple

EDIT: I am trying to achieve something similar to this

```rust
fn main() {
  let ex = smol::LocalExecutor::new();

  ex.spawn(async {
    println!("1");
    tokio::time::sleep(Duration::from_secs(1)).await;
    println!("2");
  }).detach();

  // external event loop
  loop { 
    ex.try_tick();
    if ex.is_empty() {
      break;
    }
  }
}
```

**Comment by RainM** [comment]
Hi. Do you have any updates here?

**Reply by mox692** [reply]
As far as I know, there is no such an api that does runtime task polling right now.

**Comment by tmm1** [comment]
Are you using napi-rs or rolling your own integration?

EDIT: Found https://github.com/napi-rs/napi-rs/issues/2152 and https://github.com/alshdavid/napi_ext

**Comment by SyMind** [comment]
Hi. Do you have any updates here?

**Reply by Darksonn** [reply]
You can ask Noah whether they still intend to work on https://github.com/tokio-rs/tokio/issues/6994.

**Reply by alshdavid** [reply]
I've been using the reference runtime implementation from the `futures` crate - which allows me to `run_until_stalled`.

My use-case is driving Rust futures from the Nodejs event loop, sharing the same thread. It's not compatible with tokio utilities (like sleep or channels).
